src/Plutus/Model/Mock.hs

```diff
@@ -528,84 +528,83 @@ checkSingleTx ::
   ) =>
   Core.PParams era ->
   Tx ->
-  Run (Either FailReason Stat)
-checkSingleTx params (Tx extra tx) =
-  withCheckStaking $
-    withCheckRange $
-      withTxBody $ \txBody -> do
-        let tid = fromTxId $ Ledger.txid (Class.getTxBody txBody)
-        withUTxO $ \utxo ->
-          withCheckBalance utxo txBody $
-            withCheckUnits utxo txBody $ \cost -> do
-              let txSize = fromIntegral $ BS.length $ CBOR.serialize' txBody
-                  stat = Stat txSize cost
-              withCheckTxLimits stat $ do
-                applyTx stat tid (Tx extra tx)
-                pure $ Right stat
+  Validate Stat
+checkSingleTx params (Tx extra tx) = do
+  checkStaking
+  checkRange
+  txBody <- getTxBody
+  let tid = fromTxId $ Ledger.txid (Class.getTxBody txBody)
+  utxo <- getUTxO tx
+  checkBalance utxo txBody
+  cost <- checkUnits utxo txBody
+  let txSize = fromIntegral $ BS.length $ CBOR.serialize' txBody
+      stat = Stat txSize cost
+  checkTxLimits stat
+  Validate . lift $ applyTx stat tid (Tx extra tx)
+  pure stat
   where
     pkhs = M.keys $ P.txSignatures tx
 
-    withTxBody cont = do
+    getTxBody :: Validate (Core.Tx era)
+    getTxBody = do
       cfg <- gets mockConfig
       let localScriptMap = P.txScripts tx
-      case Class.toCardanoTx localScriptMap (mockConfigNetworkId cfg) params (Tx extra tx) of
-        Right txBody -> cont txBody
-        Left err -> leftFail $ GenericFail err
+      either
+        (throwError . GenericFail)
+        pure
+        (Class.toCardanoTx localScriptMap (mockConfigNetworkId cfg) params (Tx extra tx))
 
-    withCheckStaking cont = withCheckWithdraw (withCheckCertificates cont)
+    checkStaking = do
+      checkWithdraws (extra'withdraws extra)
+      checkCertificates (extra'certificates extra)
 
-    withCheckWithdraw cont = maybe cont leftFail =<< checkWithdraws (extra'withdraws extra)
-    withCheckCertificates cont = maybe cont leftFail =<< checkCertificates (extra'certificates extra)
-
-    withCheckRange cont = do
+    checkRange :: Validate ()
+    checkRange = do
       curSlot <- gets mockCurrentSlot
-      if Interval.member curSlot $ P.txValidRange tx
-        then cont
-        else leftFail $ TxInvalidRange curSlot (P.txValidRange tx)
+      when
+        (Interval.member curSlot $ P.txValidRange tx)
+        (throwError $ TxInvalidRange curSlot (P.txValidRange tx))
```

Comment:
Doesn't look right. Shouldn't the condition be flipped?

Reply:
Oops.